In [34]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px

In [123]:
man_salary = pd.read_csv("males.csv")

In [125]:
man_salary

rownames     nr  year  school  exper union   ethn maried health  \
0            1     13  1980      14      1    no  other     no     no   
1            2     13  1981      14      2   yes  other     no     no   
2            3     13  1982      14      3    no  other     no     no   
3            4     13  1983      14      4    no  other     no     no   
4            5     13  1984      14      5    no  other     no     no   
...        ...    ...   ...     ...    ...   ...    ...    ...    ...   
4355      4356  12548  1983       9      8    no  other    yes     no   
4356      4357  12548  1984       9      9   yes  other    yes     no   
4357      4358  12548  1985       9     10    no  other    yes     no   
4358      4359  12548  1986       9     11   yes  other    yes     no   
4359      4360  12548  1987       9     12   yes  other    yes     no   

          wage                          industry  \
0     1.197540       Business_and_Repair_Service   
1     1.853060                  Personal_Service   
2     1.344462       Business_and_Repair_Service   
3     1.433213       Business_and_Repair_Service   
4     1.568125                  Personal_Service   
...        ...                               ...   
4355  1.591879                      Construction   
4356  1.212543                      Construction   
4357  1.765962                      Construction   
4358  1.745894  Professional_and_Related Service   
4359  1.466543             Public_Administration   

                          occupation   residence  
0                    Service_Workers  north_east  
1                    Service_Workers  north_east  
2                    Service_Workers  north_east  
3                    Service_Workers  north_east  
4     Craftsmen, Foremen_and_kindred  north_east  
...                              ...         ...  
4355  Craftsmen, Foremen_and_kindred         NaN  
4356  Craftsmen, Foremen_and_kindred         NaN  
4357  Craftsmen, Foremen_and_kindred         NaN  
4358  Craftsmen, Foremen_and_kindred         NaN  
4359  Craftsmen, Foremen_and_kindred         NaN  

[4360 rows x 13 columns]

In [127]:
man_salary.describe()

rownames            nr         year       school        exper  \
count  4360.000000   4360.000000  4360.000000  4360.000000  4360.000000   
mean   2180.500000   5262.058716  1983.500000    11.766972     6.514679   
std    1258.767916   3496.149815     2.291551     1.746181     2.825873   
min       1.000000     13.000000  1980.000000     3.000000     0.000000   
25%    1090.750000   2329.000000  1981.750000    11.000000     4.000000   
50%    2180.500000   4569.000000  1983.500000    12.000000     6.000000   
75%    3270.250000   8406.000000  1985.250000    12.000000     9.000000   
max    4360.000000  12548.000000  1987.000000    16.000000    18.000000   

              wage  
count  4360.000000  
mean      1.649147  
std       0.532609  
min      -3.579079  
25%       1.350717  
50%       1.671143  
75%       1.991086  
max       4.051860

In [128]:
man_salary["nr"].describe()

count     4360.000000
mean      5262.058716
std       3496.149815
min         13.000000
25%       2329.000000
50%       4569.000000
75%       8406.000000
max      12548.000000
Name: nr, dtype: float64

In [131]:
man_salary.isnull().sum()

rownames         0
nr               0
year             0
school           0
exper            0
union            0
ethn             0
maried           0
health           0
wage             0
industry         0
occupation       0
residence     1245
dtype: int64

In [133]:
man_salary["industry"].unique()

array(['Business_and_Repair_Service', 'Personal_Service', 'Trade',
       'Construction', 'Manufacturing', 'Transportation',
       'Professional_and_Related Service', 'Finance', 'Entertainment',
       'Public_Administration', 'Agricultural', 'Mining'], dtype=object)

In [135]:
man_salary["residence"].unique()

array(['north_east', nan, 'south', 'nothern_central', 'rural_area'],
      dtype=object)

### Como a base tem 4360 registros, residence tem muitos registros faltando tiraremos essa coluna da analise

In [138]:
man_salary = man_salary.drop('residence',axis = 1)
man_salary = man_salary.drop('nr',axis = 1) # I dont understand this atribute
man_salary = man_salary.drop('year',axis = 1) # not useful in the moment

In [140]:
man_salary["ethn"].describe()

count      4360
unique        3
top       other
freq       3176
Name: ethn, dtype: object

In [142]:
man_salary = man_salary.drop('ethn',axis = 1)

In [144]:
man_salary

rownames  school  exper union maried health      wage  \
0            1      14      1    no     no     no  1.197540   
1            2      14      2   yes     no     no  1.853060   
2            3      14      3    no     no     no  1.344462   
3            4      14      4    no     no     no  1.433213   
4            5      14      5    no     no     no  1.568125   
...        ...     ...    ...   ...    ...    ...       ...   
4355      4356       9      8    no    yes     no  1.591879   
4356      4357       9      9   yes    yes     no  1.212543   
4357      4358       9     10    no    yes     no  1.765962   
4358      4359       9     11   yes    yes     no  1.745894   
4359      4360       9     12   yes    yes     no  1.466543   

                              industry                      occupation  
0          Business_and_Repair_Service                 Service_Workers  
1                     Personal_Service                 Service_Workers  
2          Business_and_Repair_Service                 Service_Workers  
3          Business_and_Repair_Service                 Service_Workers  
4                     Personal_Service  Craftsmen, Foremen_and_kindred  
...                                ...                             ...  
4355                      Construction  Craftsmen, Foremen_and_kindred  
4356                      Construction  Craftsmen, Foremen_and_kindred  
4357                      Construction  Craftsmen, Foremen_and_kindred  
4358  Professional_and_Related Service  Craftsmen, Foremen_and_kindred  
4359             Public_Administration  Craftsmen, Foremen_and_kindred  

[4360 rows x 9 columns]

## para tentar prever o salário, dividiremos em X e Y nosso dataset

In [147]:
man_salaryX = man_salary.loc[:,['school','exper', 'union',"maried","health","industry","occupation"]].values

In [149]:
man_salaryX

array([[14, 1, 'no', ..., 'no', 'Business_and_Repair_Service',
        'Service_Workers'],
       [14, 2, 'yes', ..., 'no', 'Personal_Service', 'Service_Workers'],
       [14, 3, 'no', ..., 'no', 'Business_and_Repair_Service',
        'Service_Workers'],
       ...,
       [9, 10, 'no', ..., 'no', 'Construction',
        'Craftsmen, Foremen_and_kindred'],
       [9, 11, 'yes', ..., 'no', 'Professional_and_Related Service',
        'Craftsmen, Foremen_and_kindred'],
       [9, 12, 'yes', ..., 'no', 'Public_Administration',
        'Craftsmen, Foremen_and_kindred']], dtype=object)

In [151]:
man_salaryY = man_salary.loc[:, ["wage"]].values

In [153]:
man_salaryY

array([[1.1975402 ],
       [1.85306   ],
       [1.34446168],
       ...,
       [1.76596178],
       [1.74589421],
       [1.46654289]])

## devemos etratar os atributos categóricos utilizando LabelEncoder

In [156]:
from sklearn.preprocessing import LabelEncoder
label_encoder_teste = LabelEncoder()

In [158]:
man_salaryX[1,:]

array([14, 2, 'yes', 'no', 'no', 'Personal_Service', 'Service_Workers'],
      dtype=object)

In [160]:
label_encoder_union = LabelEncoder()
label_encoder_maried = LabelEncoder()
label_encoder_health = LabelEncoder()
label_encoder_industry = LabelEncoder()
label_encoder_occupation = LabelEncoder()

In [162]:
man_salaryX[:,2] = label_encoder_union.fit_transform(man_salaryX[:,2])
man_salaryX[:,3] = label_encoder_maried.fit_transform(man_salaryX[:,3])
man_salaryX[:,4] = label_encoder_health.fit_transform(man_salaryX[:,4])
man_salaryX[:,5] = label_encoder_industry.fit_transform(man_salaryX[:,5])
man_salaryX[:,6] = label_encoder_occupation.fit_transform(man_salaryX[:,6])

In [164]:
man_salaryX[1,:]

array([14, 2, 1, 0, 0, 7, 8], dtype=object)

## agora para fazer o OneHotEncoder, serve para não dar "vantagem" para nenhuma das váriaveis categóricas

In [169]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

In [171]:
onehotencoder_salary = ColumnTransformer(transformers=[('OneHot', OneHotEncoder(), [2,3,4,5,6])], remainder='passthrough')

In [173]:
man_salaryX = onehotencoder_salary.fit_transform(man_salaryX).toarray()

In [177]:
man_salaryX

array([[ 1.,  0.,  1., ...,  1., 14.,  1.],
       [ 0.,  1.,  1., ...,  1., 14.,  2.],
       [ 1.,  0.,  1., ...,  1., 14.,  3.],
       ...,
       [ 1.,  0.,  0., ...,  0.,  9., 10.],
       [ 0.,  1.,  0., ...,  0.,  9., 11.],
       [ 0.,  1.,  0., ...,  0.,  9., 12.]])

In [179]:
man_salaryX.shape

(4360, 29)

# Escalonar

In [184]:
from sklearn.preprocessing import StandardScaler
scaler_salary = StandardScaler()
man_salaryX = scaler_salary.fit_transform(man_salaryX)

In [186]:
man_salaryX[0]

array([ 0.56816858, -0.56816858,  0.88459146, -0.88459146,  0.13139831,
       -0.13139831, -0.18214103,  3.48886879, -0.2847474 , -0.12397695,
       -0.19581231, -0.62722917, -0.12587069, -0.13049224, -0.28756187,
       -0.20448963, -0.60526262, -0.26495523, -0.35419166, -0.52213134,
       -0.12205564, -0.31825808, -0.31738327, -0.5032233 , -0.34050806,
       -0.23760782,  2.75060272,  1.27895276, -1.95171915])

# Divisão das bases em treinamento e teste

In [189]:
from sklearn.model_selection import train_test_split

In [191]:
X_salary_treinamento, X_salary_teste, y_salary_treinamento, y_salary_teste = train_test_split(man_salaryX, man_salaryY, test_size = 0.2, random_state = 0)

In [193]:
X_salary_treinamento.shape

(3488, 29)

In [195]:
X_salary_teste.shape

(872, 29)

In [203]:
y_salary_treinamento.shape

(3488, 1)

In [205]:
import pickle # salvar em pickle

In [209]:
with open('salary.pkl', mode = 'wb') as f:
  pickle.dump([X_salary_treinamento, X_salary_teste, y_salary_treinamento, y_salary_teste], f)

### regressão linear


In [234]:
from sklearn.linear_model import LinearRegression
regressor_multiplo_salary = LinearRegression()
with open('salary.pkl', 'rb') as f:
 X_salary_treinamento, X_salary_teste, y_salary_treinamento, y_salary_teste = pickle.load(f)

In [236]:
regressor_multiplo_salary.fit(X_salary_treinamento,y_salary_treinamento)

LinearRegression()

In [240]:
regressor_multiplo_salary.intercept_

array([1.64887297])

In [244]:
regressor_multiplo_salary.coef_

array([[ 2.69950629e+09,  2.69950629e+09,  2.91305177e+11,
         2.91305177e+11, -8.97293191e+11, -8.97293191e+11,
        -2.14280918e+11, -3.21940701e+11, -3.20148331e+11,
        -1.48411051e+11, -2.29218217e+11, -5.47135712e+11,
        -1.50607836e+11, -1.55955762e+11, -3.22831709e+11,
        -2.38577789e+11, -5.38435780e+11, -3.00923970e+11,
        -1.39750979e+12, -1.82189789e+12, -5.34046459e+11,
        -1.28328126e+12, -1.28040035e+12, -1.78308516e+12,
        -1.35496390e+12, -9.98739235e+11, -1.42594432e+12,
         1.39897744e-01,  1.10230108e-01]])

In [246]:
len(regressor_multiplo_salary.coef_)

1

In [250]:
regressor_multiplo_salary.score(X_salary_treinamento,y_salary_treinamento)

0.25217244945415007

In [252]:
regressor_multiplo_salary.score(X_salary_teste, y_salary_teste)

0.2630361734016169

In [256]:
previsoes = regressor_multiplo_salary.predict(X_salary_teste)
previsoes

array([[1.61664641],
       [1.31171477],
       [1.58808196],
       [1.30109465],
       [1.88349211],
       [2.14472258],
       [1.44086516],
       [1.85199797],
       [1.91230071],
       [1.78461516],
       [1.28705657],
       [1.72821868],
       [1.86200774],
       [2.04877532],
       [1.3282553 ],
       [1.89496672],
       [1.20063079],
       [1.70258391],
       [1.54847014],
       [1.34369719],
       [1.62836516],
       [1.77729094],
       [2.00556243],
       [1.13117278],
       [1.56000579],
       [2.17865813],
       [1.85883391],
       [1.69696868],
       [1.72162688],
       [1.1855551 ],
       [1.87983001],
       [1.66840422],
       [1.82758391],
       [1.62854827],
       [1.30670989],
       [1.84027922],
       [1.71357024],
       [1.77191985],
       [2.11908782],
       [1.55548918],
       [1.74018157],
       [1.5827719 ],
       [1.95526946],
       [1.25232756],
       [1.89496672],
       [1.49604094],
       [1.2680136 ],
       [1.703

In [258]:
y_salary_teste

array([[ 1.39174921],
       [ 1.31507054],
       [ 2.40562352],
       [ 1.86107892],
       [ 1.35451385],
       [ 2.0800825 ],
       [ 1.34042911],
       [ 1.76032476],
       [ 1.92357539],
       [ 1.4980324 ],
       [ 1.44799957],
       [ 1.97313122],
       [ 0.59301229],
       [ 2.47540904],
       [ 1.66893317],
       [ 1.42149118],
       [ 1.03377934],
       [ 1.24213738],
       [ 1.66733189],
       [ 1.11100292],
       [ 1.81537043],
       [ 1.83300343],
       [ 2.67980655],
       [ 0.28935013],
       [ 1.76034009],
       [ 2.08856241],
       [ 1.82050194],
       [ 1.78404598],
       [ 1.81957123],
       [ 0.67558796],
       [ 1.78837022],
       [ 0.90964619],
       [ 2.68227803],
       [ 1.543916  ],
       [ 1.71036383],
       [ 1.37592288],
       [ 1.83412069],
       [ 1.93641254],
       [ 2.29909016],
       [ 1.1185408 ],
       [ 1.96383866],
       [ 1.81855529],
       [ 2.00126681],
       [ 1.59290861],
       [ 2.54725574],
       [ 1

In [260]:
from sklearn.metrics import mean_absolute_error
mean_absolute_error(y_salary_teste, previsoes)

0.32153598611789924